packages/cli/utils/gen.ts

```diff
@@ -0,0 +1,81 @@
+import { camelCase, kebabCase, pascalCase, snakeCase } from "change-case";
+
+// Keep in sync with Bucket main repo
+export const KeyFormats = [
+  "custom",
+  "pascalCase",
+  "camelCase",
+  "snakeCaseUpper",
+  "snakeCaseLower",
+  "kebabCaseUpper",
+  "kebabCaseLower",
+] as const;
+
+export type KeyFormat = (typeof KeyFormats)[number];
+
+type KeyFormatPattern = {
+  transform: (key: string) => string;
+  validate: (key: string) => true | string;
+};
+
+export const KeyFormatPatterns: Record<KeyFormat, KeyFormatPattern> = {
+  custom: {
+    transform: (key) => key?.trim(),
+    validate: (key) =>
+      /^[\p{L}\p{N}\p{P}\p{S}\p{Z}]+$/u.test(key) ||
+      "Key must contain only letters, numbers, punctuation, symbols, or spaces",
+  },
+  pascalCase: {
+    transform: (key) => pascalCase(key),
+    validate: (key) =>
+      /^[\p{Lu}][\p{L}\p{N}]*$/u.test(key) ||
+      "Key must start with uppercase letter and contain only letters and numbers",
+  },
+  camelCase: {
+    transform: (key) => camelCase(key),
+    validate: (key) =>
+      /^[\p{Ll}][\p{L}\p{N}]*$/u.test(key) ||
+      "Key must start with lowercase letter and contain only letters and numbers",
+  },
+  snakeCaseUpper: {
+    transform: (key) => snakeCase(key).toUpperCase(),
+    validate: (key) =>
+      /^[\p{Lu}][\p{Lu}\p{N}]*(?:_[\p{Lu}\p{N}]+)*$/u.test(key) ||
+      "Key must be uppercase with words separated by underscores",
+  },
+  snakeCaseLower: {
+    transform: (key) => snakeCase(key).toLowerCase(),
+    validate: (key) =>
+      /^[\p{Ll}][\p{Ll}\p{N}]*(?:_[\p{Ll}\p{N}]+)*$/u.test(key) ||
+      "Key must be lowercase with words separated by underscores",
+  },
+  kebabCaseUpper: {
+    transform: (key) => kebabCase(key).toUpperCase(),
+    validate: (key) =>
+      /^[\p{Lu}][\p{Lu}\p{N}]*(?:-[\p{Lu}\p{N}]+)*$/u.test(key) ||
+      "Key must be uppercase with words separated by hyphens",
+  },
+  kebabCaseLower: {
+    transform: (key) => kebabCase(key).toLowerCase(),
+    validate: (key) =>
+      /^[\p{Ll}][\p{Ll}\p{N}]*(?:-[\p{Ll}\p{N}]+)*$/u.test(key) ||
+      "Key must be lowercase with words separated by hyphens",
+  },
+};
+
+export function genFeatureKey(input: string, format: KeyFormat): string {
+  return KeyFormatPatterns[format].transform(input);
+}
+
+export const genDTS = (keys: string[]) => {
+  return /* ts */ `
+// DO NOT EDIT THIS FILE. IT IS GENERATED BY THE BUCKET CLI AND WILL BE OVERWRITTEN.
+// eslint-disable
+// prettier-ignore
+declare module "@bucketco/react-sdk" {
+  interface Features {
+${keys.map((key) => `    "${key}": boolean;`).join("\n")}
```

Comment:
I think this'll break if you have keys with spaces. It would also be great if we could ensure that this is well formatted so that prettier/linting wont complain. I spent some time on that here: https://github.com/bucketco/bucket-javascript-sdk/blob/feat/cli-minimal/packages/cli/utils/gen.ts#L26-L55

and wrote some tests here: https://github.com/bucketco/bucket-javascript-sdk/blob/feat/cli-minimal/packages/cli/utils/gen.test.ts

Reply:
1. I think this'll break if you have keys with spaces.
    - The quotes around keys should allow spaces.
2. It would also be great if we could ensure that this is well formatted so that prettier/linting wont complain.
    - I agree but every customer has different lint and prettier rules so this is an impossibility. We should attempt to format it nicely and then disable the linters like I have done.
3. and wrote some tests here:
    - Planning on writing tests in the next PR as I need to scaffold the whole testing suite which is a lot of files.

Reply:
https://github.com/bucketco/bucket-javascript-sdk/pull/345#issuecomment-2693924558